### PR TITLE
Support multiple epoch iteration in pretrain script

### PR DESCRIPTION
**Pull request by awaelchli**
Adds support for iterating through multiple epochs in `pretrain/tinyllama.py`.

### CODE REVIEW

File: lit_gpt/utils.py [hunk 1]
```diff
@@ -347,3 +359,33 @@ def estimate_flops(model: "GPT", training: bool) -> int:
...
+class CycleIterator:
+    """An iterator that cycles through an iterable indefinitely.
+
+    Example:
+        >>> iterator = CycleIterator([1, 2, 3])
+        >>> [next(iterator) for _ in range(5)]
+        [1, 2, 3, 1, 2]
+
+    Note:
+        Unlike ``itertools.cycle``, this iterator does not cache the values of the iterable.
```
Comment (carmocca) [inline lit_gpt/utils.py:373]: TIL

File: lit_gpt/utils.py [hunk 2]
```diff
@@ -347,3 +359,33 @@ def estimate_flops(model: "GPT", training: bool) -> int:
...
+    """
+    def __init__(self, iterable: Iterable) -> None:
+        self.iterable = iterable
+        self.epoch = 0
+        self._iterator = None
+
+    def __next__(self) -> Any:
+        if self._iterator is None:
+            self._iterator = iter(self.iterable)
```
Comment (carmocca) [inline lit_gpt/utils.py:382]: Shouldn't you do this in `__iter__`?

The current way means the iterator cannot be recreated

File: pretrain/tinyllama.py [hunk 3]
```diff
@@ -222,7 +222,7 @@ def train(fabric, state, train_dataloader, val_dataloader, resume):
             }
 
             fabric.print(
-                f"iter {metrics['iter']} step {metrics['step']}: loss {metrics['loss']:.4f}, iter time:"
+                f"iter {metrics['iter']} | epoch {train_iterator.epoch} | step {metrics['step']}: loss {metrics['loss']:.4f}, iter time:"
```
Comment (carmocca) [inline pretrain/tinyllama.py:225]: Since epochs are very long, it would be less noisy to only print the epoch when it changes instead of on every loss print.

Feel free to ignore if you disagre.

**Review by carmocca — APPROVED** [review]

File: lit_gpt/utils.py (hunk 2 above)
Comment (awaelchli) [reply lit_gpt/utils.py:382]: By it's design, I want to be able to directly call `next(CycleIterator())` without calling iter first. The `__iter__` method is just there for completeness.

Comment (awaelchli) [reply lit_gpt/utils.py:382]: To recreated a second such iterator, you would do `CycleIterator(iterable)` again.

In this sense, `CycleIterator` can be seen as a pure replacement for `iter`.

File: pretrain/tinyllama.py (hunk 3 above)
Comment (awaelchli) [reply pretrain/tinyllama.py:225]: On the other hand, if the epochs are very long, you'll need to scroll the log very long to find the epoch print.

Alternatively, I could remove it from the print and log it instead if you prefer that more.

Comment (carmocca) [reply pretrain/tinyllama.py:225]: I like that suggestion. Usually the terminal width is very limited
